### DISCUSSION TITLE
How could I add space around content?

### DISCUSSION THREAD

**Discussion by chhpt** [microsoft/monaco-editor#3815] [Q&A]
Because of some reasons, I want to add some space around the content(view-lines). I found the way to add space on the left of the content using the API lineDecorationsWidth, but I have no idea how to add space on the other side, top, bottom and right and keep the action work well, such as copy, select all, input...

**Comment by chhpt** [comment]
I had asked this question on the [stackoverflow](https://stackoverflow.com/questions/50054544/monaco-editor-add-space-around-content), but no one have answer me. 😔

**Comment by myfoxtail** [comment]
@chhpt may be you're looking for the `lineHeight`?

**Comment by chhpt** [comment]
@myfoxtail Could the `lineHeight` change the space on the left or right? I can't find the usage that show it could change the space around the content. It only change the space between the line. I don't think the `lineHeight` is the proper way.

**Comment by myfoxtail** [comment]
@chhpt I guess you can try to create a decoration and apply it to the particular range if your content. Every decoration has it's own css class which I guess you can fill with margins and paddings.

```
const decorationId = this.editor.deltaDecorations([], [
        {
          range: new monaco.Range(
              lineNumberStart
            , columnStart
            , lineNumberEnd
            , columnEnd),
          options: {
            inlineClassName: 'myClass', // use this class to add some paddings
            stickiness     : 1,
            hoverMessage   : {value: 'hello my friend'}
          }
        }
      ]);
```

**Comment by chhpt** [comment]
@myfoxtail  I have tried it. It seems that it is not working. The `inlineClassName` has effect on the inline block. But when I add padding on it, it cause the content chaos. And I also have tried the `className`, it is not work as expect. May the API didn't fit this situation.

**Comment by alexdima** [comment]
You could try with decorations that use `::before` and `::after` css selectors, like e.g. `beforeContentClassName`

```js
var jsCode = [
	'"use strict";',
	'function Person(age) {',
	'	if (age) {',
	'		this.age = age;',
	'	}',
	'}',
	'Person.prototype.getAge = function () {',
	'	return this.age;',
	'};'
].join('\n');

var editor = monaco.editor.create(document.getElementById("container"), {
	value: jsCode,
	language: "javascript"
});

var decorations = editor.deltaDecorations([], [
	{ range: new monaco.Range(7,1,7,24), options: { beforeContentClassName: 'myInlineDecoration' }},
]);

```

```css
.myInlineDecoration::before {
    content: " ";
    background: red;
    width: 100px;
    display: inline-block;
}
```

**Comment by chhpt** [comment]
@alexandrudima  

Maybe I should show my thoughts more clearly. First please see this pic. If I don't do anything, it look like this. This doesn't look so good. 

![](http://test-1252710547.cosgz.myqcloud.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202018-04-30%20%E4%B8%8B%E5%8D%8816.11.46%20%E4%B8%8B%E5%8D%88.png)

So I decided to add some space around the content. I add a `padding` attribute on the  parent container, it became like this:

![](http://test-1252710547.cosgz.myqcloud.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202018-04-30%20%E4%B8%8B%E5%8D%8816.00.44%20%E4%B8%8B%E5%8D%88.png)

All is well except the scrollbar. I want to find way that could make it work like this but make the scrollbar on the edge. May like this:

![](http://test-1252710547.cosgz.myqcloud.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202018-04-30%20%E4%B8%8B%E5%8D%8816.10.13%20%E4%B8%8B%E5%8D%88.png)

The `beforeContentClassName` API is something else, it make cause like this. it is not what I except.
And if edit the content, such as delete something, it work weird. Is there a way could change the position of the scrollabr?

```JavaScript
var decorations = editor.deltaDecorations([], [
    { range: new monaco.Range(1,1,10,24), options: { beforeContentClassName: 'myInlineDecoration' }},
]);
```

![](http://test-1252710547.cosgz.myqcloud.com/%E5%B1%8F%E5%B9%95%E5%BF%AB%E7%85%A7%202018-04-30%20%E4%B8%8B%E5%8D%8816.23.41%20%E4%B8%8B%E5%8D%88.png)

**Comment by alexdima** [comment]
Thanks for posting screenshots, it makes it clearer for me what you're trying to do. 

At this time, the editor doesn't support any form of padding inside the content (such that the scrollbar stays put).

**Comment by chhpt** [comment]
@alexandrudima  Do you have a plan to realize it? I hope the feature could be come ture.

**Comment by Mani-108** [comment]
Any update related this feature? @alexdima

**Comment by JesseCHN** [comment]
Any update related this feature? @alexdima

**Comment by diyyap** [comment]
@alexdima How do I get to put a consistent single space between words on the same line. I am using a plaintext editor. Please help me . I want to use create a data like tab separated files. Eg. 
Name    address
Johnd    london
Dhdhb   Newyork

**Comment by knackstedt** [comment]
I would also like to ask for this functionality :)

This provides a very useful inlay ability for consumers to inject content. I believe this should be fairly possible similar to how the "peek" mode has been implemented -- as it would effectively work in the same way.

**Comment by knackstedt** [comment]
I ended up using a sort of workaround by using ViewZones. In my case, I also want it to show the text that's selected, so it includes a small sample implementation of that.

Visually, it inserts a block "decoration" fairly well. I hadn't encountered any major problems so far.

![image](https://github.com/microsoft/monaco-editor/assets/29349167/99ff6b8c-e32e-431b-b931-713cbea33373)


[Playground link](https://microsoft.github.io/monaco-editor/playground.html?source=v0.44.0#XQAAAAIbEwAAAAAAAABBqQkHQ5NjdMjwa-jY7SIQ9S7DNlzs5W-mwj0fe1ZCDRFc9ws9XQE0SJE1jc2VKxhaLFIw9vEWSxW3yscxV6ODYuecutyzd5agpCNWq_EI3LBWy-5jucEA64F_PqhVsw2RWL36WkDr_RlUvVeNPM0qyBuSz9HW5wkci9LopKHNdmATfpuLsdC2-ur-c01eTqat2Ke56HbzNJgtzhmU7KDUYXYe1FKWyjqp5nSY2i185aVdz8KUGoofZmlyXJdLTGwP0tLol6cQTHyFXhZOX4Ko1hHqOfcep5RCXNgip0-fjiuATnAYI46_Bb0Z7hONfAZcvOIXWYybdUa36Re-b1PrOfw80XQ9-tVX0r_QqN3Ykz3pXUIZGG6EflrSc2ZnARABd7uxP5tIv_BWWbwGVl1-UMKFMBVlSeoPi5hApHn4CauPVvnQf48AeMYODtJRcdA32gxE1c4JvQ0sE-nv-lc2OenK087ZBzP3sT0VSFxsy2rTPU92MPW52T33NKOdc1osmONjT3ThePI3fOwIM7pp7ON41y7k5nk3nBxGHznFrH1ohKTxR_V8EUN-NWV3mBHnOH7DIUC7FuPnoQpYzBySiJ7h6xxEFLXLhplPr_VnUU8d3w_4XkTEYNByXBwv-xdXodYP6iMBmHs73MWeTUOm5hAcX1r7uPTxYyfrwVkCuJzR-AHeXrFo4Pcb-1Eg_rIeAmID0gDG3G2ajwlliZx_ZQZAjRVJc3Yf8wVd0dDG6MRCZU8SV3paCnCalsii-lRy0Lysrp3jf7G7DVisI5FhqfmUhd_lPdFK9IDXVW0-QZsQRG8O1n0PETU2ptwHXccRPz-dau7cBmlvCKg7RCqo8KR6ADWBzRwELKjAXXG386UnGD-JlN7ZU7VH693xhD0ttBRPp5ARGy3AdPe7Jg0gFAiO2jmRPmX9SbFHWDqXkPD6HFdTqjXSmhCFvq9X7IcpsKFf_OrP2vXCQuQGjT-jdYEXji3tJ-LXJlXIvEzL4HPw4D8ogysTPHoc01ywT3iMihJx-aMrcor_sitbN7CtxTi0YrwOmA7dypcH5Q71RdBM8d2lGjKJv842PC696gww3WEEpSmNVucOiMP3Hr6yvPgnHHzIckcDMZvyL8phpM6ZnV651RuzBymjaiKxkxjpEXcusMJcv9sl-QwDTMszRrOLTS8Uy4Qm7PHJe5CiBzoR6LDlvD5s2qXzKwwsFVlFlUrOXODnvtNvgz1HPzcC0fDem9f29cw6HksMp-T3pjku5MjYU9M7KMCtAFfi1LMwJijVPErsWqy1qLbe-QdA1LgS0cI6B9O-EZxj0okSTwCwv_ARsH-pc_4lF3sXk2ZrPZzmQ-4L4w3A0l8Ze_VWwuz7uY1r5I3pIX6l216c5rOIPvHNgt22JRiwIwbTUdCgbNGX1llFrNsOsQs6RQ6tJewVCXYuGuMvISqB20SslF1Cxvso7U0l1gCRXkwKAdmVyGMalg9y-Nl9BqmzQzrIis3vqO5hCtFW8DS8jAFhqpI8a5tb0pVqLXiFlyka4tIPVrOptIMz36FMFv9dRzVVwXsZA2BYUjdUB2HbO0rYTYy4OAHx0Cjg8M2F8yLZDL5rd41JQ1KrsJEsutGpD1jue-bnaYVr0xp0RyZMDtON9QMI5Lc1f-t0cMQD-kkJQPDQHf_l_EyLk0cu8zzEZzvFR-8HxteARfjkogpOQ-b3QL5SYeqUv2fSmNr_qMFw2zphC4RHjg5qbCyQV_V8NLe497VB4W-P8La5EYQvaO2V4NAf4q2M3QXE8RilVnlEu47RkoLtChi1JvnkNv2uEGdRiIn8gOj9BTk_vba3tJX4n1JNzbR2TI_SLizOPbdVo0m4yesKHTgahTchiOhTN7FTtU40L3udGG_tixdp6BgRevulknRdTX8MA3Sqop9c0JsCu9ePJtePrbs_vAwBZJ3LdlKR2MPbQq11Zr9p7XLMqlSiren_2g5GEd6dhBZVmjbZ4DkdqmLLtdSEomJTKi1we6MXnHtdXxOj7UCIDbkaw_4TsDDey6NcCM5OC7pEi6EX7GNbh-q1osUoL4Os_L_hO6LRmJKF4V0WSx6l92iVrLbgPe96SBeYUjxoWWK7eiGJVj-82CNvfnSZgxQvwUXys0fYzh0zdHmFFF1UoT4YYlLo8PE8vwEb05bwvsDaH32OEe3IMi6BM7hTPCl0LqqsBCFFQ78P1lk99aCqJEhGRdnxi29h2zu1SY0eFyzfF31etBe8Pjm_py9tV3vF9ixDv4Dpvk9A18KymxsXfSFmUXm0bYfHRL87gDYtKRuT8h4iEvt_iQoAxEh5A2handbQoCNnvR_8oLdclyTdqhnAoM4FK59XZ1O6w8QYwnk1wUlVyOgbKH_at2IiFEqtn5fz-bD_FfXNG6QM8184Ff_9PkbW)
